skimage/segmentation/trainable_segmentation.py

```diff
@@ -0,0 +1,180 @@
+
+        Raises
+        ------
+        NotFittedError if ``self.clf`` has not been fitted yet (use ``self.fit``).
+        """
+        if self.features_func is None:
+            self.features_func = multiscale_basic_features
+        features = self.features_func(image)
+        return predict_segmenter(features, self.clf)
+
+
+def fit_segmenter(labels, features, clf):
+    """
+    Segmentation using labeled parts of the image and a classifier.
+
+    Parameters
+    ----------
+    labels : ndarray of ints
+        Image of labels. Labels >= 1 correspond to the training set and
+        label 0 to unlabeled pixels to be segmented.
+    features : ndarray
+        Array of features, with the first dimension corresponding to the number
+        of features, and the other dimensions correspond to ``labels.shape``.
+    clf : classifier object
+        classifier object, exposing a ``fit`` and a ``predict`` method as in
+        scikit-learn's API, for example an instance of
+        ``RandomForestClassifier`` or ``LogisticRegression`` classifier.
+
+    Returns
+    -------
+    output : ndarray
+        Labeled array, built from the prediction of the classifier trained on
+        ``labels``.
+    clf : classifier object
+        classifier trained on ``labels``
+
+    Raises
+    ------
+    NotFittedError if ``self.clf`` has not been fitted yet (use ``self.fit``).
+    """
+    training_data = features[:, labels > 0].T
+    training_labels = labels[labels > 0].ravel()
+    clf.fit(training_data, training_labels)
+    data = features[:, labels == 0].T
+    predicted_labels = clf.predict(data)
+    output = np.copy(labels)
+    output[labels == 0] = predicted_labels
+    return output, clf
+
+
+def predict_segmenter(features, clf):
+    """
+    Segmentation of images using a pretrained classifier.
+
+    Parameters
+    ----------
+    features : ndarray
+        Array of features, with the first dimension corresponding to the number
+        of features, and the other dimensions are compatible with the shape of
+        the image to segment.
+    clf : classifier object
+        trained classifier object, exposing a ``predict`` method as in
+        scikit-learn's API, for example an instance of
+        ``RandomForestClassifier`` or ``LogisticRegression`` classifier. The
+        classifier must be already trained, for example with
+        :func:`skimage.segmentation.fit_segmenter`.
+    features_func : function, optional
+        function computing features on all pixels of the image, to be passed
+        to the classifier. The output should be of shape
+        ``(m_features, *labels.shape)``. If None,
+        :func:`skimage.segmentation.multiscale_basic_features` is used.
+
+    Returns
+    -------
+    output : ndarray
+        Labeled array, built from the prediction of the classifier.
+    """
+    sh = features.shape
+    features = features.reshape((sh[0], np.prod(sh[1:]))).T
```

Comment:
@emmanuelle why did you choose to put the features in front and not at the end? This is not cache friendly at all! Or does sklearn prefer F-ordered arrays?

Also, note that you can reshape to `(sh0, -1)`. =)

Reply:
hum, indeed not cache-friendly. While building the array of features it made sense to compute an array for each feature and to stack them together, hence features at the front. But I should probably use `np.ascontiguousarray` here.

Reply:
or do you think it would make more sense to ask that the features array is of shape `img.shape + (n_features,)`? This is possible too.

Reply:
@emmanuelle I guess you can stack directly with the features in the tail dimension, no? I'm unclear whether we should go for `(npixels, nfeatures)` or `image.shape + (nfeatures,)`. I have a slight leaning towards the former — let's maximise compatibility with sklearn, as I think that is the purpose of this module.

Reply:
In `predict_segmenter` you need to have `image.shape` in the shape of `features` in order to be able to reshape the labels to the correct image shape. Actually it could be possible to support several shapes, and just assume that the last dimensions is for the features. This is what I'll do.